Refactor immersive mode implementation for hang prevention and loading animations

MortimerGoro opened the pull request:
* Fixes #306, #285, #263, #259, #257, #122 
* Depends on https://bugzilla.mozilla.org/show_bug.cgi?id=1481393 for a correct exit from WebVR in all the situations
* Link traversal doesn't work because of this: https://bugzilla.mozilla.org/show_bug.cgi?id=1481475
* The spinner is temporal, we'll add a rotation 3D model when ready

Some pages that can be used for testing:

- Presentation with no Submit Frames:
  - https://webvr.info/samples/test-slow-render.html?noFrames=1
  - https://virtualart.chromeexperiments.com/vr/

- Slow EnterVR 
  - https://developer.playcanvas.com/en/tutorials/webvr-lab/

- Presentation that stops submitting frames after some time (good to test the frame discard code on each platform)
  
  -[03-vr-presentation.html.zip](https://github.com/MozillaReality/FirefoxReality/files/2267103/03-vr-presentation.html.zip)

bluemarvin commented on app/src/main/cpp/BrowserWorld.cpp:
```diff
@@ -793,37 +807,54 @@ BrowserWorld::DrawWorld() {
 #endif // !defined(VRBROWSER_NO_VR_API)
-  m.device->EndFrame();
+
+   m.device->EndFrame(false);
```
Extra space on indent?

bluemarvin commented on app/src/main/cpp/ExternalVR.cpp:
```diff
@@ -128,17 +128,23 @@ struct ExternalVR::State {
+    pthread_mutex_init(&data.systemMutex, nullptr);
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
```
I would prefer we avoid recursive locks (http://www.fieryrobot.com/blog/2008/10/14/recursive-locks-will-kill-you/)

Why is this needed?

bluemarvin commented:
I really don't like the spinning wait animation. Maybe something that pulses would be better? A spinning  image in the dark is really uncomfortable in VR (for me). @thenadj

MortimerGoro commented:
> I think we need to re-think the locking strategy. I think we are overloading the mutexes. I think we might need four mutexes and two condition variable. One mutex to protect the browser data, one to protect the system state, one plus condition variable to block browser side, and one pause condition variable to block on system side. Using the same mutex to both protect data and block seems to be causing a lot of unnecessary issues. I realize this is more difficult because we need to change the Gecko side too.

mmm I think that adding more mutexes could add some race conditions. For example what would happen if Gecko pushes some change to the browser data in the interval between we Pull the browser data (exit condition not satisfied) and before locking using the mutex for the condition variable? I think that we'd miss the awake. That doesn't happen now because the mutex is released atomically when the condition variable starts the lock.

Other alternative would be to  create a private function (e.g.) called PullBrowserStateInternal() that does the logic that we don't want to duplicate in the code. We are only want to reuse some code with checks after pulling the browser data while not presenting and  when we are using the condition variable in a WebVR frame. This would be a variation of the lock argument we had before 

`I really don't like the spinning wait animation. Maybe something that pulses would be better? A spinning image in the dark is really uncomfortable in VR (for me). @thenadj`

I agree. I like the one in the Oculus browser better.

BTW, I had to add some transparent area on right & bottom of the PNG to make the center of the image the same as the center of the spiral. But it's still confusing.

bluemarvin commented:
The race condition is a good concern. I wonder if we should hide VRExternalShmem. Maybe instead of looking at VRExternalShmem directly, we spawn a thread that does a condition var wait on the VRBrowserState. Every time VRBrowserState is changed, the thread wakes up and figures out what has changed and passes that to the VR render thread as an event? Then the render thread does a timed wait on events. If there is no event, we can show the waiting animation?

bluemarvin commented on app/src/main/cpp/LoadingAnimation.cpp:
```diff
@@ -0,0 +1,93 @@
+
+
+void
+LoadingAnimation::Update() {
```
This reminds me, I want to update vrb::Updatable::UpdateResource() to get passed the delta time from last update so it can be used for animation nodes as well. https://github.com/MozillaReality/vrb/issues/34

bluemarvin commented:
The remainder of the code looks good. But I would really like to see if we can figure out a way to simply  how we interface with VRExternalShmem so the locking isn't so complicated.

MortimerGoro commented:
My concern with the event thread is that we may end up in the same situation to check/pull the events: mutex + wait locks, needing the same lock pattern.

I updated the PR removing the PTHREAD_MUTEX_RECURSIVE attribute by using the internal function I mentioned in the prev comment. I think it's a good pattern, the blogpost you shared recommends it too (http://www.fieryrobot.com/blog/2008/10/14/recursive-locks-will-kill-you). Also added some comments to make it easier to understand.

We can merge this for now and continue the locking conversation and the loading animation delta times as a follow-ups.

bluemarvin commented:
After sleeping on it, a separate thread might be overkill. The problem is that we can already miss things since nothing is queued. By having a dedicated thread we can create an event pump back to the render thread so we decrease the chance that we miss a state change in the VRExternalShmem.

avrignaud commented:
Note: once spinner model is animated (3D model), we should also remove the small spinner in the URL bar.

bluemarvin commented:
@andrenatal I think the spinner 3D model currently will only show when going into immersive mode. The spinner in the URL bar is probably a different issue?
